src/seq/demultiplexer.jl

```diff
@@ -0,0 +1,336 @@
+            error("duplicated sequences")
+        end
+    end
+    nodes = Int64[1]
+    build_trie!(nodes, 1, barcodes, ids, 1:endof(barcodes), 1)
+    return BarcodeTrie(nodes)
+end
+
+# Build a trie recursively.
+function build_trie!(nodes, root, barcodes, ids, range, j)
+    n = length(range)
+    if n == 0
+        return
+    elseif n == 1
+        i = first(range)
+        if j > length(barcodes[i]) + 1
+            # leaf
+            nodes[root] = ids[i] << 1
+            return
+        end
+    end
+
+    # allocate five consecutive nodes {'#', 'A', 'C', 'G', 'T'}
+    resize!(nodes, length(nodes) + 5)
+    for l in endof(nodes)-4:endof(nodes)
+        nodes[l] = 0
+    end
+    base = endof(nodes) - 5
+
+    # calculate ranges for the next level
+    starts = Vector{Int}(5)
+    i = first(range)
+    starts[1] = i
+    while i ≤ last(range) && length(barcodes[i]) < j
+        i += 1
+    end
+    for k in 1:4
+        starts[k+1] = i
+        while i ≤ last(range) && Int(barcodes[i][j]) + 1 == k
+            i += 1
+        end
+    end
+
+    # set nodes
+    nodes[root] = (base << 1) | 1
+    for k in 1:5
+        if k == 5
+            r = starts[k]:last(range)
+        else
+            r = starts[k]:starts[k+1]-1
+        end
+        build_trie!(nodes, base + k, barcodes, ids, r, j + 1)
+    end
+end
+
+# Find the longest match of `seq` prefixes in `trie`.
+function findbarcode(trie::BarcodeTrie, seq)
+    s = 1
+    for nt in seq
+        t = (trie.nodes[s] >> 1) + min(Int(nt), 3) + 2
+        if trie.nodes[t] & 1 == 0
+            # leaf
+            break
+        end
+        s = t
+    end
+    s = (trie.nodes[s] >> 1) + 1
+    return trie.nodes[s] >> 1
+end
+
+type Demultiplexer{S}
+    # barcode tries with ≥ 0 errors in ascending order
+    # (i.e. tries[1]: no errors, tries[2]: one error, ...)
+    tries::Vector{BarcodeTrie}
+
+    # original barcodes
+    barcodes::Vector{S}
+
+    # distance (:hamming or :levenshtein)
+    distance::Symbol
+end
+
+function Base.show(io::IO, demultiplexer::Demultiplexer)
+    println(io, summary(demultiplexer), ":")
+    println(io, "  distance: ", demultiplexer.distance)
+    println(io, "  number of barcodes:", length(demultiplexer.barcodes))
+      print(io, "  number of correctable errors: ", length(demultiplexer.tries) - 1)
+end
+
+"""
+    Demultiplexer(barcodes::Vector{DNASequence};
+                  n_max_errors::Integer=1,
+                  distance::Symbol=:hamming)
+
+Create a demultiplexer object from `barcodes`.
+
+# Arguments
+* `barcodes`: the sorted list of DNA barcodes; integer IDs are assigned in this order.
+* `n_max_errors=1`: the number of maximum correctable errors.
+* `distance=:hamming`: the distance metric (`:hamming` or `:levenshtein`).
+"""
+function Demultiplexer(barcodes::Vector{DNASequence};
+                       n_max_errors::Integer=1,
+                       distance::Symbol=:hamming)
+    if !issorted(barcodes)
```

Comment:
Thinking about it more carefully, I found there is no need to check the order of barcodes. So I removed it.